File: leetcode/leetcode_p143/code_143.cpp
// 力扣第143题，重排链表
/*
给定一个单链表 L：L0→L1→…→Ln-1→Ln ，
将其重新排列后变为： L0→Ln→L1→Ln-1→L2→Ln-2→…

你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
*/
#include <iostream>
#include <deque>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(): val(0), next(nullptr){}
    ListNode(int x): val(x), next(nullptr){}
    ListNode(int x, ListNode *next): val(x), next(next){}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head){
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast->next != nullptr && fast->next->next != nullptr)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }

    ListNode* reverseList(ListNode* head){
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (curr != nullptr)
        {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    void mergeList(ListNode* l1, ListNode* l2){
        ListNode* l1_tmp;
        ListNode* l2_tmp;
        while(l1 != nullptr && l2 != nullptr){
            l1_tmp = l1->next;
            l2_tmp = l2->next;

            l1->next = l2;
            l1 = l1_tmp;

            l2->next = l1;
            l2 = l2_tmp;
        }
    }

    void reorderList(ListNode* head){
        if (head == nullptr){
            return;
        }
        ListNode* mid = middleNode(head);
        ListNode* l1 = head;
        ListNode* l2 = mid->next;
        mid->next = nullptr;
        l2 = reverseList(l2);
        mergeList(l1, l2);
    }
};


int main()
{
    cout << "find middle + list_reverse + merge_list" << endl;

    cin.get();
    return 0;
}

File: leetcode/leetcode_p143/code_143.py
# 力扣第143题，重排链表
"""
给定一个单链表 L：L0→L1→…→Ln-1→Ln ，
将其重新排列后变为： L0→Ln→L1→Ln-1→L2→Ln-2→…

你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
"""
import collections

class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head):
        if not head:
            return 
        record = []
        cur = head
        while cur:
            record.append(cur)
            cur = cur.next

        cur = record.pop(0)
        while record:
            cur.next = record.pop()
            cur = cur.next
            if len(record) == 0:
                break
            cur.next = record.pop(0)
            cur = cur.next
        cur.next = None


class Solution:
    def reorderList(self, head):
        if not head:
            return 
        record = collections.deque()
        cur = head
        while cur:
            record.append(cur)
            cur = cur.next

        cur = record.popleft()
        while record:
            cur.next = record.pop()
            cur = cur.next
            if len(record) == 0:
                break
            cur.next = record.popleft()
            cur = cur.next
        cur.next = None


if __name__ == "__main__":
    pass